### DISCUSSION TITLE
The Latency of the First Frame is Too High

### DISCUSSION THREAD

**Discussion by AvatarGanymede** [microsoft/msquic#2089] [Q&A]
Hi, sorry to interrupt again.  😌
I'm working on a project and the work flow is as below:
```
Server: Captures desktrop and Sends it through Quic Stream
Client: 
            case QUIC_STREAM_EVENT_RECEIVE:
                    copy QUIC Buffer and Push the copied data for the process on the other thread
```
I've known that MsQuic internally seperates my captured frame into a few small packets, so I add a header flag ahead of the frame data:
```
---------------------------------------
|| header | frame || header | frame ||
---------------------------------------
```
Thus I know where the frame data ends on client side and assemble these small packets into a complete frame.
Then I try to test the time cost of each frame, from receiving the first packet to the last packet:

![image](https://user-images.githubusercontent.com/59817181/138252038-6a02749c-71a1-4663-a34a-c712d41fd4a8.png)

It's obvious that, the length of these frame data is close, but the latency of the first frame is way too high.

```
StreamOpen(Connection, QUIC_STREAM_OPEN_FLAG_0_RTT, ServerStreamCallback, this, &m_stream);
...
StreamStart(m_stream, QUIC_STREAM_START_FLAG_ASYNC | QUIC_STREAM_START_FLAG_IMMEDIATE);
...
StreamSend(m_stream, m_sendBuffer, 1, QUIC_SEND_FLAG_ALLOW_0_RTT, m_sendBuffer)
```

Do I Miss some crutial setting or using something wrong?

**Comment by nibanks** [comment]
Is this first payload the first one the whole connection? I suspect the connection handshake is what is taking so long the first time around.

**Reply by AvatarGanymede** [reply]
It is the first one in the whole connection.  How can I know when the handshake ends? or how can I send valid data after handshake?

**Reply by AvatarGanymede** [reply]
I send the first payload after the Event `QUIC_CONNECTION_EVENT_CONNECTED` is triggered. In the doc, this Event means that the handshake has completed.

**Reply by nibanks** [reply]
If you're sending it after that, then the handshake should be complete. Can you [grab logs](https://github.com/microsoft/msquic/blob/main/docs/Diagnostics.md#trace-collection) and share them out?

**Reply by AvatarGanymede** [reply]
[log.zip](https://github.com/microsoft/msquic/files/7394061/log.zip)
Server is sending side and Client is receiveing side.
The 'QUIC Excution' Window of the Client is as below:
![image](https://user-images.githubusercontent.com/59817181/138387819-362c2447-f4ce-4505-a709-108db899beae.png)
No.19 TimeAckDelay seems quite time-consuming.

**Reply by nibanks** [reply]
Ah, so you're saying the delayed ACK might be involved here? Exactly when do you stop your time measurement, for the data above? When you receive the FIN / peer shutdown notification? Or when you get shutdown complete?

**Reply by AvatarGanymede** [reply]
I measure the time, from I receive the header of the frame, to the end of the frame. All the thing is done after the client receives `QUIC_STREAM_EVENT_RECEIVE`. 
The stream never shuts donw until the last frame has been sent.

**Reply by chunlei-wu** [reply]
> I send the first payload after the Event `QUIC_CONNECTION_EVENT_CONNECTED` is triggered. In the doc, this Event means that the handshake has completed.

Hi AatarGanymede,

Do you find the root cause and solution now? I want to learn it, many thanks.
